lib/code-caller-docker.js

```diff
@@ -0,0 +1,685 @@
+    }
+}
+
+/**
+ * 
+ * @param {string} mountpoint 
+ */
+async function removeBindMount(mountpoint) {
+    if (process.platform === 'darwin') {
+        await spawnAsync('umount', [mountpoint]);
+    } else if (process.platform === 'linux') {
+        await spawnAsync('fusermount', ['-u', mountpoint]);
+    } else {
+        throw new Error(`Unsupported platform for bind mounts: ${process.platform}`);
+    }
+}
+
+class DockerCaller {
+    constructor() {
+        debug('enter constructor()');
+
+        this.uuid = uuidv4();
+        this.container = null;
+        this.callback = null;
+        this.timeoutID = null;
+        this.ensureChildMutex = new Mutex();
+        this.ensureDirectoryMutex = new Mutex();
+
+        // this will accumulate stdout from the container
+        this.output = '';
+
+        // for error logging
+        this.lastCallData = null;
+
+        this.course = null;
+
+        /** @type {CallerState} */
+        this.state = CREATED;
+        this._checkState();
+
+        debug(`exit constructor(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+    }
+
+    /**
+     * Allows this caller to prepare for execution of code from a particular
+     * course.
+     *
+     * @param {{ path: string }} course 
+     * @param {(err: Error | null | undefined) => void} callback 
+     */
+    prepareForCourse(course, callback) {
+        util.callbackify(async () => {
+            if (!this._checkState([CREATED, WAITING])) throw new Error(`Invalid caller state: ${String(this.state)}`);
+            if (this.course) {
+                if (this.course.path === course.path) {
+                    // Same course as before; we can reuse the existing setup
+                    return;
+                } else if (config.workersExecutionMode === 'container') {
+                    throw new Error(`DockerCaller is already allocated for course "${this.course.path}"`);
+                }
+            }
+
+            this.course = course;
+
+            if (config.workersExecutionMode === 'container') {
+                // If we're running in Docker, that means we're making use of static
+                // docker mounts instead of messing with bind mounts on the host.
+
+                if (this.container) {
+                    // We already have a container proeprly allocated for this course
+                    return;
+                }
+
+                const elementsHostDir = path.join(process.env.HOSTFILES_DIR, 'elements');
+                const pythonHostDir = path.join(process.env.HOSTFILES_DIR, 'python');
+                let courseDir;
+                if (course.path === '/PrairieLearn/exampleCourse') {
+                    // TODO: can we get away without hardcoding this? I don't think so, but...
+                    courseDir = path.join(process.env.HOSTFILES_DIR, 'exampleCourse');
+                } else {
+                    // TODO: make this better? maybe?
```

Comment:
I agree it feels a bit nasty, but I can’t really see a great alternative?

Reply:
I ripped out all this code in the end 🎉